evm/forge/tests/gas/CircleIntegrationComparison.t.sol

```diff
@@ -0,0 +1,440 @@
+
+    function setupWormhole() public {
+        wormhole = IWormhole(vm.envAddress("TESTING_WORMHOLE_ADDRESS"));
+    }
+
+    function setupUSDC() public {
+        (, bytes memory queriedDecimals) =
+            USDC_ADDRESS.staticcall(abi.encodeWithSignature("decimals()"));
+        uint8 decimals = abi.decode(queriedDecimals, (uint8));
+        assertEq(decimals, 6, "wrong USDC");
+    }
+
+    function setupCircleIntegration() public {
+        // deploy Setup
+        Setup setup = new Setup();
+
+        // deploy Implementation
+        Implementation implementation = new Implementation(
+            address(wormhole),
+            uint8(1), // finality
+            vm.envAddress("TESTING_CIRCLE_BRIDGE_ADDRESS") // tokenMessenger
+        );
+
+        // deploy Proxy
+        ERC1967Proxy proxy =
+            new ERC1967Proxy(address(setup), abi.encodeCall(setup.setup, address(implementation)));
+
+        circleIntegration = ICircleIntegration(address(proxy));
+        circleIntegration.setUpOverride(uint256(vm.envBytes32("TESTING_DEVNET_GUARDIAN")));
+
+        forkedCircleIntegration = ICircleIntegration(FORKED_CIRCLE_INTEGRATION_ADDRESS);
+    }
+
+    function setUp() public {
+        // set up circle contracts (transferring ownership to address(this), etc)
+        setupUSDC();
+
+        // set up wormhole simulator
+        setupWormhole();
+
+        // now our contract
+        setupCircleIntegration();
+
+        composedContract =
+            new ComposingWithCircleIntegration(address(circleIntegration), USDC_ADDRESS);
+
+        inheritedContract = new InheritingWormholeCctp(
+            address(wormhole),
+            address(circleIntegration.circleBridge()), // tokenMessenger
+            USDC_ADDRESS
+        );
+    }
+
+    function test_Inherited__TransferUsdc(uint256 amount, bytes32 mintRecipient, bytes32 data)
+        public
+    {
+        amount = bound(amount, 1, _cctpBurnLimit());
+        vm.assume(mintRecipient != bytes32(0));
+        vm.assume(data != bytes32(0));
+
+        (uint16 targetChain,) = _registerEmitterAndDomain(circleIntegration);
+        bytes memory payload = _generatePayload512(data);
+
+        _dealAndApproveUsdc(amount, inheritedContract);
+
+        inheritedContract.transferUsdc(amount, mintRecipient, payload);
+
+        // This is here to avoid the unused variable warning.
+        targetChain;
+    }
+
+    function test_Composed__TransferUsdc(uint256 amount, bytes32 mintRecipient, bytes32 data)
+        public
+    {
+        amount = bound(amount, 1, _cctpBurnLimit());
+        vm.assume(mintRecipient != bytes32(0));
+        vm.assume(data != bytes32(0));
+
+        (uint16 targetChain,) = _registerEmitterAndDomain(circleIntegration);
+        bytes memory payload = _generatePayload512(data);
+
+        _dealAndApproveUsdc(amount, composedContract);
+
+        composedContract.transferUsdc(targetChain, amount, mintRecipient, payload);
```

Comment:
It might be nice to add balance checks before and after transferring the USDC here.